File: hw1/runtime-test.js
let mergeSort = require('./merge-sort-func');
let insertSort = require('./insert-sort-func');

executeTests('merge', 'sorted');
executeTests('merge', 'random');
executeTests('merge', 'reversed');
executeTests('insert', 'sorted');
executeTests('insert', 'random');
executeTests('insert', 'reversed');

function executeTests(algorithm, useCase) {
    let title = `${algorithm.toUpperCase()} RUNTIME TEST. NUMBERS IN THE ARRAY ARE ${useCase.toUpperCase()}.`;
    console.log(title);
    let func;

    switch (algorithm) {
        case 'merge':
            func = mergeSort;
            break;
        case 'insert':
            func = insertSort;
            break;
    }

    for (let i = 1; i <= 10; i++) {
        test(func, i * 10000, useCase);
    }
}

function test(algorithm, qty, useCase) {
    let name = 'Number of elements: ' + qty;
    let arr;


    switch (useCase) {
        case 'sorted':
            arr = generateSortedArray(qty);
            break;
        case 'reversed':
            arr = generateReversedArray(qty);
            break;
        default:
            arr = generateRandomNumbers(qty);
    }

    console.time(name);
    algorithm(arr);
    console.timeEnd(name);
}


function generateRandomNumbers(qty) {
    let arr = [];

    for (let i = 0; i < qty; i++) {
        arr.push(randomNumber(10000));
    }

    return arr;
}

function generateSortedArray(length) {
    let arr = [];

    for (let i = 0; i < length; i++) {
        arr.push(i);
    }

    return arr;
}

function generateReversedArray(length) {
    let arr = [];

    for (let i = length - 1; i >=0; i--) {
        arr.push(i);
    }

    return arr;
}

function randomNumber(max) {
    let random = Math.random() * Math.floor(max);
    return Math.floor(random);
}

File: hw1/python/runtime-test.py
import random
import time
from merge_sort_func import merge_sort
from insert_sort_func import insert_sort


def run_performance_tests(algorithm, use_case):
    for i in range(1, 8):
        test(algorithm, i * 5000, use_case)


def test(algorithm, qty, use_case):
    arr = []

    if use_case == "sorted":
        arr = generate_sorted_array(qty)
    elif use_case == "reversed":
        arr = generate_reversed_array(qty)
    elif use_case == "worst_merge":
        arr = worstCaseArrayOfSize(qty)
    else:
        arr = generate_random_numbers(qty)

    start = time.time()
    if algorithm == "merge":
        merge_sort(arr)
    elif algorithm == "insert":
        insert_sort(arr)
    end = time.time()
    exec_time = end - start
    # get ms
    exec_time *= 1000

    log = ""
    log += algorithm.upper()
    log += " SORT. ARRAY IS "
    log += use_case.upper()
    log += ". QTY: "
    log += str(qty)
    log += ". TIME: "
    log += str(round(exec_time, 2))
    print(log)


def generate_random_numbers(qty):
    arr = []
    for i in range(qty):
        arr.append(random_number(10000))
    return arr


def generate_sorted_array(length):
    arr = []
    for i in range(length):
        arr.append(i)
    return arr


def generate_reversed_array(length):
    arr = []
    for i in range(length, 0, -1):
        arr.append(i)
    return arr


def random_number(max_num):
    return random.randint(1, max_num + 1)


print("=========MERGE SORT TESTS=========")
run_performance_tests('merge', 'sorted')
run_performance_tests('merge', 'random')
run_performance_tests('merge', 'random')
run_performance_tests('merge', 'worst_merge')

print("=========INSERT SORT TESTS=========")
run_performance_tests('insert', 'sorted')
run_performance_tests('insert', 'random')
run_performance_tests('insert', 'reversed')
